docs/CIP_Id_filters.md

```diff
@@ -0,0 +1,64 @@
+# CIP-5: Id Filters Proposal
+
+## Status
+
+Current Status: `Under Discussion`
+
+## **Motivation**
+
+Currently, Chroma does not provide a way to pre-filter embeddings with Id values. Doing so would enable users to perform pre-query processes for more complex and/or efficient queries. This has been requested by [community members](https://discord.com/channels/1073293645303795742/1074711446589542552/1088185430035411006).
+
+## **Public Interfaces**
+
+The changes will affect the following public interfaces:
+
+- `collection.query()`
+
+## **Proposed Changes**
+
+We suggest the introduction of a new `where_id` argument to `query()`.
```

Comment:
I'm concerned with adding still more fields to `query`, it would be better if we could unify this under the same filter structure (and bring `where_document` in as well)

Reply:
I suppose we could combine all of them into `where` and use something like this:

```python
collection.query(query_texts=query,
                 where={
                    # current `where` functionality
                    "metadata": {
                       "tag": { "$eq": "some_tag"}
                    },
                    # proposed `where_id` functionality
                    "id": {
                       "$in": [1, 2, 3]
                    },
                    # current `where_document` functionality
                    "document": {
                       "$contains":"search_string"
                    }
                 },
                 n_results=3)
```

What do you think @atroyn?

Reply:
@tyherox, this is nice, but we should consider all the breakage this will cause for the existing install base. Let's consider several options and pick the one that makes the most sense for the community.

1. existing public interface - `where` to be replaced with the syntax above
2. new param `filter` that can encapsulate the above-suggested changes
3. ??

Reply:
This PR's priority is adding in the `id` filter so going through with the first option looks the best atm @tazarov.

If we want to introduce this feature as seemless as possible we should go with the original proposition of adding a new `where_id` param. I know @atroyn was worried about adding more params but maybe we can address combinging the different fitlers in a separate PR that can be included in the next major release so we can properly warn people of the breakage.

Reply:
@tyherox, alright let's assume we take the step to introduce a new parameter as you suggested originally but with a twist - add the where_id to a filter param as I suggested in option 2. Then you end up with the small change you intend but you encapsulate it in a future-proof way, e.g. next release we can move more parameters into the filter.

Does this sound reasonable?

Reply:
Cool, sounds like a plan. So if i understand this correctly, you're looking for something like this:

```python
collection.query(query_texts=query,
                 where={"metadata_field": "is_equal_to_this"},
                 where_document={"$contains":"search_string"}
                 filter={
                    "where_id": {
                       "$in": [1, 2, 3]
                    },
                 },
                 n_results=3)
```

Looking good?

Reply:
yes, that is exactly what I meant. @atroyn wdyt about this as an approach?

Reply:
Bump on this guys - I'll get started once a decision is made.

Reply:
@atroyn any thoughts on this approach?

Reply:
I think we should just skip ahead to the breaking change in this PR, indicate when it's coming ahead of time, and support people who encounter it and have problems. 

Doing it half-way feels like just kicking the can down the road, and at least some users will have the problem that they were using both the new `filter` variable *and* the old `where` filters when we do make the change. 

@HammadB happy to hear alternative perspectives here but I'm more happy to make breaking changes now so that more users have a better experience in the future, as long as we communicate those breaking changes effectively, present a clear path for migrating existing code, and support people running into the problem.

Reply:
@atroyn, does it make sense to go with half-way and every time someone uses any of the `where`, `where_document` etc, filters, Chroma prints a deprecation warning? This will allow time for people to adjust and nudge them (via warning messages) into the new way of filtering. Under the hood, we should refactor code to transform legacy filters into the new approach with a single filter so that we can seamlessly deprecate stuff with minimal refactoring.

Alternatively, if this is not a priority feature, push it to 0.5.x alongside distributed Chroma.

Reply:
We can do the LC refactors for them, the integration is not very complex, and we can coordinate the releases appropriately. 

I think we should push the change before distributed so it's not too much complexity in a single release all at the same time. 

We could add an exception that tells people what change they need to make in the first release with this change, if they still have those arguments set with the new client.

Reply:
I don't have an alternate perspective. I am fine with doing whatever the "correct" thing is but I haven't seen a good proposal for the "correct" thing. Special keywords in the map feels wrong.

Reply:
So by the look of things we will just bite the bullet with this PR and move forward with the breaking changes with warnings to notify users of the breaking change. This will also require a change in the langchain repositories. 

And to be clear, is my previous proposal for a unifed interface a good to go from your end @atryon? 

```python
collection.query(query_texts=query,
                 where={
                    # current `where` functionality
                    "metadata": {
                       "tag": { "$eq": "some_tag"}
                    },
                    # proposed `where_id` functionality
                    "id": {
                       "$in": [1, 2, 3]
                    },
                    # current `where_document` functionality
                    "document": {
                       "$contains":"search_string"
                    }
                 },
                 n_results=3)
```